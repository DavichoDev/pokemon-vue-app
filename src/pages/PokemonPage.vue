<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>

  <div v-else>

    <h1>¿Quién es este pokémon?</h1>

    <PokemonImage 
      :pokemonId="pokemon.id" 
      :showPokemon="showPokemon" />

    <PokemonOptions 
      :pokemons="pokemonArr" 
      @selected-pokemon="checkAnswer"
    />

    <template v-if="message"
         class="fade-in">
      <h2>{{ message }}</h2>
      <button @click="newGame">Nuevo juego</button>
    </template>


  </div>

</template>

<script>

import PokemonImage from '@/components/PokemonImage.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'

import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
  components: {
    PokemonImage,
    PokemonOptions
  },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      message: null
    }
  },
  methods: {
    async mixedPokemonArray() {

      this.pokemonArr = await getPokemonOptions()
      
      const rndInt = Math.floor( Math.random() * 4 )

      this.pokemon = this.pokemonArr[ rndInt ]

    },
    checkAnswer( selectedId ) {

      if (selectedId === this.pokemon.id) {
        this.message = `Correcto, ${ this.pokemon.name }`
      } else { 
        this.message = `Oops, era ${ this.pokemon.name }`
      }

      this.showPokemon = true

    },
    newGame() {

      this.message = null
      this.showPokemon = false,
      this.pokemon = null
      this.pokemonArr = []
      this.mixedPokemonArray()

    }
  },
  mounted() {
    this.mixedPokemonArray()
  },
}
</script>

<style>

</style>