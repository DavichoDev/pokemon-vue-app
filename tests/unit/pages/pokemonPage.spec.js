import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { pokemonsMock } from '../mocks/pokemons.mock'
// Componentes
import PokemonImage from '@/components/PokemonImage.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'

describe('PokemonPage Component', () => {

    let wrapper

    beforeEach(() => {
      wrapper = shallowMount( PokemonPage )

      jest.clearAllMocks()
    })

    test('Debe hacer match con el Snapshot', () => {
        expect( wrapper.html() ).toMatchSnapshot()
    })
    
    test('Debe de llamar mixedPokemonArray al montar', () => {
        const mixedPokemonArray = jest.spyOn( PokemonPage.methods, 'mixedPokemonArray' )
        const wrapper = shallowMount( PokemonPage )

        expect( mixedPokemonArray ).toHaveBeenCalled()
    })

    test('Debe de hacer match con el snapshot cuando cargan los pokemon', () => {

      const wrapper = shallowMount( PokemonPage, {
        data() {
          return {
            pokemonArr: pokemonsMock,
            pokemon: pokemonsMock[0],
            showPokemon: false,
            message: null
          } 
        }
      })

      expect( wrapper.html() ).toMatchSnapshot( )

    })

    test('Debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {
      
      const wrapper = mount( PokemonPage, {
        data() {
          return {
            pokemonArr: pokemonsMock,
            pokemon: pokemonsMock[0],
            showPokemon: true,
            meesage: null
          }
        }
      })
      const pokemonImage = wrapper.findComponent(PokemonImage)
      const pokemonOptions = wrapper.findComponent(PokemonOptions)

      expect( pokemonImage.exists() ).toBeTruthy()
      expect( pokemonOptions.exists() ).toBeTruthy()
      expect( pokemonImage.props('pokemonId') ).toBe(1)
      expect( pokemonOptions.props('pokemons') ).toBeTruthy()

    })

    test('Prubas con checkAnswer', async() => {
      const wrapper = shallowMount( PokemonPage, {
        data() {
          return {
            pokemonArr: pokemonsMock,
            pokemon: pokemonsMock[0],
            showPokemon: false,
            meesage: null
          }
        }
      })

      
      // Pokemon correcto
      await wrapper.vm.checkAnswer(1)
      expect( wrapper.find('h2').text() ).toBe(`Correcto, ${ pokemonsMock[0].name }`)
      expect( wrapper.vm.message ).toBe(`Correcto, ${ pokemonsMock[0].name }`)

      // Revisa si el pokemon es visible
      expect( wrapper.vm.showPokemon ).toBeTruthy()
      
      // Revisa si el mensaje es visible
      expect( wrapper.find('h2').exists() ).toBeTruthy()

      // Pokemon incorrecto
      await wrapper.vm.checkAnswer(2)
      expect( wrapper.find('h2').text() ).toBe(`Oops, era ${ pokemonsMock[0].name }`)
      expect( wrapper.vm.message ).toBe(`Oops, era ${ pokemonsMock[0].name }`)
      
    })

})