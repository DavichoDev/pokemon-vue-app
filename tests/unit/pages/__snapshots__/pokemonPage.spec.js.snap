// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component Debe de hacer match con el snapshot cuando cargan los pokemon 1`] = `
<div>
  <h1>¿Quién es este pokémon?</h1>
  <pokemon-image-stub pokemonid="1" showpokemon="false"></pokemon-image-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;

exports[`PokemonPage Component Debe hacer match con el Snapshot 1`] = `<h1>Espere por favor...</h1>`;
